File: evals/fixtures/stale_latest_state/go/safe_all_paths_write.go
package session

// Both branches update s.latest (an explicit empty result on failure), so a
// reader always reflects the most recent Run.
func (s *Session) Run(q string) (Result, error) {
	r, err := execute(q)
	if err != nil {
		s.latest = Result{Empty: true}
		return r, err
	}
	s.latest = r
	return r, nil
}

func (s *Session) RepeatLatest(p Plan) Result {
	return apply(p, s.latest)
}

File: evals/fixtures/stale_latest_state/python/safe_all_paths_write.py
_latest = None


def record_snapshot(columns, rows):
    global _latest
    _latest = {"columns": columns, "rows": rows}


def get_latest_snapshot():
    return _latest


def run_query(sql):
    columns, rows = execute(sql)
    # Every producing path writes the store, including the empty-result case,
    # so the reader always reflects the latest turn. No staleness.
    record_snapshot(columns, rows)
    return render_table(columns, rows)


def reformat_last(plan):
    snap = get_latest_snapshot()
    return apply(plan, snap["rows"])
